src/nodes/hypertable_modify.c

```diff
@@ -93,6 +95,287 @@ get_chunk_dispatch_states(PlanState *substate)
+ */
+static bool
+is_segmentby_col_null(Chunk *curr_chunk, char *column_name, int *attrno)
+{
+	Chunk *comp_chunk = NULL;
+	Relation chunk_rel;
+	Relation comp_chunk_rel;
+	TupleDesc tupleDesc;
+	int attno = 0;
+	bool is_null = false;
+
+	chunk_rel = relation_open(curr_chunk->table_id, AccessShareLock);
+	tupleDesc = RelationGetDescr(chunk_rel);
+	/*
+	 * check if SEGMENT BY column is NULLABLE or not based on uncompressed
+	 * chunk tuple descriptor. This is because compressed chunk column
+	 * attributes don't have NULLABLE properties.
+	 */
+	for (attno = 1; attno <= tupleDesc->natts; attno++)
+	{
+		Form_pg_attribute attribute = TupleDescAttr(tupleDesc, AttrNumberGetAttrOffset(attno));
+		if (attribute->attisdropped)
+			continue;
+		char *attributeName = NameStr(attribute->attname);
+		if (strcmp(attributeName, column_name) == 0)
+		{
+			*attrno = attno;
+			is_null = attribute->attnotnull;
+			break;
+		}
+	}
+	table_close(chunk_rel, AccessShareLock);
+	/*
+	 * we get attrno for SEGMENT BY column from uncompressed chunk. This
+	 * can be different from that of compressed chunk due to different
+	 * physical layouts of compressed and uncompressed chunks. Thus we
+	 * need to retreive correct attrno from compressed chunk tuple descriptor.
+	 */
+	comp_chunk = ts_chunk_get_by_id(curr_chunk->fd.compressed_chunk_id, true);
+	comp_chunk_rel = relation_open(comp_chunk->table_id, AccessShareLock);
+	tupleDesc = RelationGetDescr(comp_chunk_rel);
+	for (attno = 1; attno <= tupleDesc->natts; attno++)
+	{
+		Form_pg_attribute attribute = TupleDescAttr(tupleDesc, AttrNumberGetAttrOffset(attno));
+		if (attribute->attisdropped)
+			continue;
+		char *attributeName = NameStr(attribute->attname);
+		if (strcmp(attributeName, column_name) == 0)
+		{
+			*attrno = attno;
+			break;
+		}
+	}
+	table_close(comp_chunk_rel, AccessShareLock);
+	return !is_null;
+}
+
+/*
+ * Helper method which returns true if given column_name
+ * is configured as SEGMENT BY column in a compressed hypertable
+ */
+static bool
+is_segmentby_col(List *ht_compression_info, char *column_name)
+{
+	ListCell *lc;
+	foreach (lc, ht_compression_info)
+	{
+		FormData_hypertable_compression *fd = lfirst(lc);
+		if (namestrcmp(&fd->attname, column_name) == 0)
+		{
+			if (fd->segmentby_column_index > 0)
+				return true;
+			break;
+		}
+	}
+	return false;
+}
+
+/*
+ * Helper method which will build scan keys for SEGMENT BY column
+ * with index attribute number from compressed chunk.
+ */
+static ScanKeyData*
+build_update_delete_scankeys(Chunk *curr_chunk, char *column_name, Const *const_value,
+							 int strategy)
+{
+	Chunk *comp_chunk = NULL;
+	Relation comp_chunk_rel;
+	TupleDesc tupleDesc;
+	int attno = 0;
+	Form_pg_attribute attr;
+	Datum value;
+	ScanKeyData *scankey = NULL;
+	TypeCacheEntry *tce;
+
+	comp_chunk = ts_chunk_get_by_id(curr_chunk->fd.compressed_chunk_id, true);
+	comp_chunk_rel = relation_open(comp_chunk->table_id, AccessShareLock);
```

Comment:
relation_open/relation_close and table_open/table_close should be used together and not mixed

Reply:
Fixed.